contributors/design-proposals/no-new-privs.md

```diff
@@ -0,0 +1,141 @@
+`no_new_privs` set, `execve` promises not to grant the privilege to do anything
+that could not have been done without the `execve` call.
+
+For more details about `no_new_privs`, please check the
+[Linux kernel documention](https://www.kernel.org/doc/Documentation/prctl/no_new_privs.txt).
+
+This is different from `NOSUID` in that `no_new_privs`can give permission to
+the container process to further restrict child processes with seccomp. This
+permission goes only one-way in that the container process can not grant more
+permissions, only further restrict.
+
+### Interactions with other Linux primitives
+
+- suid binaries: will break when `no_new_privs` is enabled
+- seccomp2 as a non root user: requires `no_new_privs`
+- seccomp2 with dropped `CAP_SYS_ADMIN`: requires `no_new_privs`
+- ambient capabilities: requires `no_new_privs`
+- selinux transitions: bugs that were fixed documented [here](https://github.com/moby/moby/issues/23981#issuecomment-233121969)
+
+
+## Current Implementations
+
+### Support in Docker
+
+Since Docker 1.11, a user can specify `--security-opt` to enable `no_new_privs`
+while creating containers, for example
+`docker run --security-opt=no_new_privs busybox`.
+
+Docker provides via their Go api an object named `ContainerCreateConfig` to
+configure container creation  parameters. In this object, there is a string
+array `HostConfig.SecurityOpt` to specify the security options. Client can
+utilize this field to specify the arguments for security options while
+creating new containers.
+
+This field did not scale well for the Docker client, so it's suggested that
+Kubernetes does not follow that design.
+
+This is not on by default in Docker.
+
+More details of the Docker implementation can be read
+[here](https://github.com/moby/moby/pull/20727) as well as the original
+discussion [here](https://github.com/moby/moby/issues/20329).
+
+### Support in rkt
+
+Since rkt v1.26.0, the `NoNewPrivileges` option has been enabled in rkt.
+
+More details of the rkt implementation can be read
+[here](https://github.com/rkt/rkt/pull/2677).
+
+### Support in OCI runtimes
+
+Since version 0.3.0 of the OCI runtime specification, a user can specify the
+`noNewPrivs` boolean flag in the configuration file.
+
+More details of the OCI implementation can be read
+[here](https://github.com/opencontainers/runtime-spec/pull/290).
+
+## Existing SecurityContext objects
+
+Kubernetes defines `SecurityContext` for `Container` and `PodSecurityContext`
+for `PodSpec`. `SecurityContext` objects define the related security options
+for Kubernetes containers, e.g. selinux options.
+
+To support "no new privileges" options in Kubernetes, it is proposed to make
+the following changes:
+
+## Changes of SecurityContext objects
+
+Add a new `*bool` type field named `allowPrivilegeEscalation` to the `SecurityContext`
+definition.
+
+By default, ie when `allowPrivilegeEscalation=nil`, we will set `no_new_privs=true`
+with the following exceptions:
```

Comment:
> confused as to what you mean

Perhaps I got confused because there are 2 fields named `allowPrivilegeEscalation` and table bellow explains the 2nd in the PSP while here we're talking about SC.

> By default, ie when `allowPrivilegeEscalation=nil`, we will set `no_new_privs=true` with the following exceptions:

In other words, if `allowPrivilegeEscalation=nil`, we will set `no_new_privs=true` when `SecurityContext.RunAsUser=nil` or `SecurityContext.RunAsUser=0`.

All other cases are exceptions and `no_new_privs` will be set to `false`.

Is this correct?